pkg/controller/apmserver/apmserver_controller.go

```diff
@@ -180,29 +178,34 @@ func (r *ReconcileApmServer) Reconcile(request reconcile.Request) (reconcile.Res
 
 	var as apmv1beta1.ApmServer
 	if ok, err := association.FetchWithAssociation(r.Client, request, &as); !ok {
-		return reconcile.Result{}, err
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		r.onDelete(types.NamespacedName{
+			Namespace: request.Namespace,
+			Name:      request.Name,
+		})
+		return reconcile.Result{}, nil
 	}
 
 	if common.IsPaused(as.ObjectMeta) {
 		log.Info("Object is paused. Skipping reconciliation", "namespace", as.Namespace, "as_name", as.Name)
 		return common.PauseRequeue, nil
 	}
 
-	if err := r.finalizers.Handle(&as, r.finalizersFor(as)...); err != nil {
-		if errors.IsConflict(err) {
-			log.V(1).Info("Conflict while handling secret watch finalizer")
-			return reconcile.Result{Requeue: true}, nil
-		}
+	if compatible, err := r.isCompatible(&as); err != nil || !compatible {
 		return reconcile.Result{}, err
 	}
 
-	if as.IsMarkedForDeletion() {
-		// APM server will be deleted nothing to do other than run finalizers
-		return reconcile.Result{}, nil
+	// Remove any previous Finalizers
+	if err := finalizer.RemoveAll(r.Client, &as); err != nil {
+		return reconcile.Result{}, err
 	}
 
-	if compatible, err := r.isCompatible(&as); err != nil || !compatible {
-		return reconcile.Result{}, err
+	if as.IsMarkedForDeletion() {
```

Comment:
I'm wondering whether that's necessary or the case where the resource is not found is enough?

Reply:
I've been wondering the same question, I think that "the resource is not found" is enough, but I'm not 100% sure.
Since the idea was also to delete Secrets asap and since it does not hurt I call the cleanup logic at both place.